[dailymotion] extract m3u8 formats and handle live feeds(fixes #6794)

remitamine opened the pull request:
there is also f4m manifest but it's difficut to get it because it's url hidden in the swf file.

jaimeMF commented on youtube_dl/extractor/dailymotion.py:
```diff
@@ -390,12 +395,19 @@ def _real_extract(self, url):
+        if is_live:
+            title = self._live_title(title)
+        else:
+            formats.append({'url': self.redirect(video_info['mp4_url'], video_id), 'format_id': 'mp4'})
```
Apparently those urls use a normal HTTP 302 redirect, since the user may not download this format we can avoid the additional request. I haven't tested the m3u8 url, but if it also uses a normal redirect we should also avoid it.

remitamine replied:
the m3u8 doesn't work if i pass the url directly because i think the m3u8 downloader does when it create the relative links of segments it uses the original url not the redirected url.
without redirect function:
```
youtube-dl -f 4401 'http://replay.publicsenat.fr/vod/documentaire/crash/169683'
[generic] 169683: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 169683: Downloading webpage
[generic] 169683: Extracting information
[DailymotionCloud] 55ed5dc794a6f61d74350384: Downloading webpage
[DailymotionCloud] 55ed5dc794a6f61d74350384: Downloading m3u8 information
[download] Destination: DOCUMENTAIRE - Crash-55ed5dc794a6f61d74350384.m3u8
ffmpeg version 2.7.2 Copyright (c) 2000-2015 the FFmpeg developers
  built with gcc 5.2.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gnutls --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libschroedinger --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-shared --enable-version3 --enable-x11grab
  libavutil      54. 27.100 / 54. 27.100
  libavcodec     56. 41.100 / 56. 41.100
  libavformat    56. 36.100 / 56. 36.100
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 16.101 /  5. 16.101
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  2.100 /  1.  2.100
  libpostproc    53.  3.100 / 53.  3.100
[http @ 0x55aea48d9ba0] HTTP error 404 NOT FOUND
http://cdn.dmcloud.net/route/hls/522497e194a6f650d300002d/55ed5dc794a6f61d74350384/abs-audio%3D151982-video%3D4000000.m3u8?e=1441745234&st=ECb4WDb-IiAxCFvSdXZwWw: Server returned 404 Not Found


ERROR: ffmpeg exited with code 1
```
with redirect function:
```
youtube-dl -f 4401 'http://replay.publicsenat.fr/vod/documentaire/crash/169683'
[generic] 169683: Requesting header
WARNING: Falling back on generic information extractor.
[generic] 169683: Downloading webpage
[generic] 169683: Extracting information
[DailymotionCloud] 55ed5dc794a6f61d74350384: Downloading webpage
[DailymotionCloud] 55ed5dc794a6f61d74350384: Downloading m3u8 information
[download] Destination: DOCUMENTAIRE - Crash-55ed5dc794a6f61d74350384.m3u8
ffmpeg version 2.7.2 Copyright (c) 2000-2015 the FFmpeg developers
  built with gcc 5.2.0 (GCC)
  configuration: --prefix=/usr --disable-debug --disable-static --disable-stripping --enable-avisynth --enable-avresample --enable-fontconfig --enable-gnutls --enable-gpl --enable-libass --enable-libbluray --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmodplug --enable-libmp3lame --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libopenjpeg --enable-libopus --enable-libpulse --enable-libschroedinger --enable-libsoxr --enable-libspeex --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-shared --enable-version3 --enable-x11grab
  libavutil      54. 27.100 / 54. 27.100
  libavcodec     56. 41.100 / 56. 41.100
  libavformat    56. 36.100 / 56. 36.100
  libavdevice    56.  4.100 / 56.  4.100
  libavfilter     5. 16.101 /  5. 16.101
  libavresample   2.  1.  0 /  2.  1.  0
  libswscale      3.  1.101 /  3.  1.101
  libswresample   1.  2.100 /  1.  2.100
  libpostproc    53.  3.100 / 53.  3.100
Input #0, hls,applehttp, from 'http://usp-05.dmcloud.net/522497e194a6f650d300002d/55ed5dc794a6f61d74350384/abs-1441617151.ism/abs-audio%3D151982-video%3D4000000.m3u8?e=1441745318&st=R20XkbGwYAhO9l4HLqKy6A':
  Duration: 01:03:16.00, start: 10.000000, bitrate: 0 kb/s
  Program 0 
    Metadata:
      variant_bitrate : 0
    Stream #0:0: Audio: aac (LC) ([15][0][0][0] / 0x000F), 44100 Hz, stereo, fltp, 145 kb/s
    Stream #0:1: Video: h264 (Constrained Baseline) ([27][0][0][0] / 0x001B), yuv420p, 1280x704 [SAR 1:1 DAR 20:11], 25 fps, 25 tbr, 90k tbn, 50 tbc
[mp4 @ 0x561b9bb8aba0] Codec for stream 0 does not use global headers but container format requires global headers
[mp4 @ 0x561b9bb8aba0] Codec for stream 1 does not use global headers but container format requires global headers
Output #0, mp4, to 'DOCUMENTAIRE - Crash-55ed5dc794a6f61d74350384.m3u8.part':
  Metadata:
    encoder         : Lavf56.36.100
    Stream #0:0: Video: h264 ([33][0][0][0] / 0x0021), yuv420p, 1280x704 [SAR 1:1 DAR 20:11], q=2-31, 25 fps, 25 tbr, 90k tbn, 90k tbc
    Stream #0:1: Audio: aac ([64][0][0][0] / 0x0040), 44100 Hz, stereo, 145 kb/s
Stream mapping:
  Stream #0:1 -> #0:0 (copy)
  Stream #0:0 -> #0:1 (copy)
Press [q] to stop, [?] for help
frame=   22 fps= 22 q=-1.0 size=    1112kB time=00:00:00.84 bitrate=10840.6kbitsframe=   22 fps=9.2 q=-1.0 size=    1183kB time=00:00:00.84 bitrate=11540.8kbits/s    
ERROR: Interrupted by user
```
it can solved with one three ways(from the best to the worst):
- make the m3u8 downloader to use redirected url not original url for relative paths of segments.
- make a direct request(HEAD) to the url and get the redirected url from the HTTP 302 redirect.
- using the redirect method(like in the js of the website).

there is another thing the ios_url doesn't always contain a `m3u8` manifest sometimes it contain a url of file inside it there is a `moov` box with direct urls of different formats, this is the output using `curl`:
�moov�rmra�rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac-1433266013.mp4?auth=1757075603-1-o6ucwtrk-61db517519fc3a3196bf3bb3d78f29b3rmdrdrmcs��rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac_hd-1433266081.mp4?auth=1757075603-1-l0xp99fz-e6e4943f9ae1234fd9812fd307cca462rmdrI�rmcs��rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac_ld-1433265989.mp4?auth=1757075603-1-l38rze7s-c9dc0dc2e07cc11e81c12ac02db68779rmdr
�rmcs,

remitamine commented:
i changed the _extract_m3u8_formats to use the redirected url for relative urls.
now m3u8 manifests works with the normal HTTP 302 redirect.

remitamine commented:
> there is another thing the ios_url doesn't always contain a m3u8 manifest sometimes it contain a url of file inside it there is a moov box with direct urls of different formats, this is an example of the output using curl:
```
��moov��rmra�rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac-1433266013.mp4?auth=1757075603-1-o6ucwtrk-61db517519fc3a3196bf3bb3d78f29b3�rmdrd�rmcs���rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac_hd-1433266081.mp4?auth=1757075603-1-l0xp99fz-e6e4943f9ae1234fd9812fd307cca462�rmdr�I��rmcs���rmda�rdrfurl �http://cdn.dmcloud.net/route/http/4e7343f894a6f677b10006b4/556e03339473995ee145930c/mp4_h264_aac_ld-1433265989.mp4?auth=1757075603-1-l38rze7s-c9dc0dc2e07cc11e81c12ac02db68779�rmdr
��rmcs�,
```

should i add a todo for parsing .mov files.

remitamine commented:
i added support for extracting f4m_url from the swf file and fixed an two errors in f4m_downloader:
the first error related to redirected url as in the _extract_m3u8_formats and handle query string properly:
before it try to download this urls:
http://usp-05.dmcloud.net/522497e194a6f650d300002d/55ed5dc794a6f61d74350384/abs-1441617151.ism/abs-audio%3D151982-video%3D4000000-?e=1441944376&st=xKfvZNsh_4wYyexOiyzauwSeg1-Frag1
and now it can get the right url:
http://usp-05.dmcloud.net/522497e194a6f650d300002d/55ed5dc794a6f61d74350384/abs-1441617151.ism/abs-audio%3D151982-video%3D4000000-Seg1-Frag1?e=1441944376&st=xKfvZNsh_4wYyexOiyzauw
